pynumdiff/kalman_smooth/_kalman_smooth.py

```diff
@@ -6,638 +6,255 @@
 
 
-    :param options: a dictionary indicating whether to run smoother forwards and backwards
-                    (usually achieves better estimate at end points)
-    :type params: dict {'forwardbackward': boolean}, optional
-
-    :return: a tuple consisting of:
-
-            - x_hat: estimated (smoothed) x
-            - dxdt_hat: estimated derivative of x
-
-
-    :rtype: tuple -> (np.array, np.array)
+def _RTS_smooth(xhat0, P0, y, A, C, Q, R, u=None, B=None)
+    """forward-backward Kalman/Rauch-Tung-Striebel smoother. For params see the helper functions.
     """
-    if len(x.shape) == 2:
-        pass
-    else:
-        x = np.reshape(x, [1, len(x)])
-
-    if options is None:
-        options = {'forwardbackward': True}
-
-    if options['forwardbackward']:
-        x_hat_f, smooth_dxdt_hat_f = __constant_velocity__(x, dt, params, options={'backward': False})
-        x_hat_b, smooth_dxdt_hat_b = __constant_velocity__(x, dt, params, options={'backward': True})
-
-        w = np.arange(0, len(x_hat_f), 1)
-        w = w/np.max(w)
-
-        x_hat = x_hat_f*w + x_hat_b*(1-w)
-        smooth_dxdt_hat = smooth_dxdt_hat_f*w + smooth_dxdt_hat_b*(1-w)
-
-        smooth_dxdt_hat_corrected = np.mean((smooth_dxdt_hat, smooth_dxdt_hat_f), axis=0)
-
-        return x_hat, smooth_dxdt_hat_corrected
-
-    return __constant_velocity__(x, dt, params, options={'backward': False})
-
-
-#########################
-# Constant Acceleration #
-#########################
+    xhat_pre, xhat_post, P_pre, P_post = _kalman_forward_filter(xhat0, P0, x, A, C, Q, R) # noisy x are the "y" in Kalman-land
+    xhat_smooth, _ = _kalman_backward_smooth(A, xhat_pre, xhat_post, P_pre, P_post) # not doing anything with P_smooth
+    return xhat_smooth
 
 
-def __constant_acceleration__(x, dt, params, options=None):
+#########################################
+# Constant 1st, 2nd, and 3rd derivative #
+#########################################
+def _constant_derivative(x, P0, A, C, R, Q, forwardbackward)
+    """Helper for `constant_{velocity,acceleration,jerk}` functions, because there was a lot of
+    repeated code.
     """
-    Run a forward-backward constant acceleration RTS Kalman smoother to estimate the derivative.
-
-    :param x: array of time series to differentiate
-    :type x: np.array (float)
-
-    :param dt: time step size
-    :type dt: float
-
-    :param params: a list of two elements:
-
-                    - r: covariance of the x noise
-                    - q: covariance of the constant velocity model
-
-    :type params: list (float)
+    xhat0 = np.zeros(A.shape[0]); xhat0[0] = x[0]
+    xhat_smooth = _RTS_smooth(xhat0, P0, x, A, C, Q, R) # noisy x are the "y" in Kalman-land  
+    x_hat_forward = xhat_smooth[:, 0] # first dimension is time, so slice first element at all times
+    dxdt_hat_forward = xhat_smooth[:, 1]
+
+    if not forwardbackward: # bound out here if not doing the same in reverse and then combining
```

Comment:
It's very odd to me that the Kalman with RTS smoothing already works forward and backward, but then you invert the dynamics and go again and then take a weighting of the answers. Why? I thought RTS smoothing was already supposed to be optimal. Did it not seem to work at one point, and these were further hacks? I wonder if having the bugs in the RTS smoother fixed will make the hacks unnecessary.

Reply:
The answer is that a non-optimal guess to kick off the process biases the end you start on, and it takes a while for that bias to wear off. Classic RTS doesn't fully address this, but doing two of them and weighting together does.

Ideally, you could choose an initial x0,P0 that wouldn't horribly bias the process, but due to $C$ not being full rank, there is no way to choose the single optimal `x0,P0`. See my ponderings on #97.